Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/FileDescriptorLogStream.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.IO;
+using System.Buffers;
+using Microsoft.Win32.SafeHandles;
+using System.Collections.Concurrent;
+
+namespace Amazon.Lambda.RuntimeSupport.Helpers
+{
+    /// <summary>
+    /// This class wraps the utility of writing to the Lambda telemetry file descriptor for logging into a standard .NET Stream.
+    /// The message
+    /// </summary>
+    public static class FileDescriptorLogFactory
+    {
+        static ConcurrentDictionary<string, StreamWriter> _writers = new ConcurrentDictionary<string, StreamWriter>();
+
+        /// <summary>
+        /// Get the StreamWriter for the particular file descriptor ID. If the same ID is passed the same StreamWriter instance is returned.
+        /// </summary>
+        /// <param name="fileDescriptorId"></param>
+        /// <returns></returns>
+        public static StreamWriter GetWriter(string fileDescriptorId)
+        {
+            // AutoFlush must be turned out otherwise the StreamWriter might not send the data to the stream before the Lambda function completes.
+            var writer = _writers.GetOrAdd(fileDescriptorId, (x) => new StreamWriter(new FileDescriptorLogStream(fileDescriptorId)) { AutoFlush = true });
+            return writer;
+        }
+
+        /// <summary>
+        /// Write log message to the file descriptor which will make sure the message is recorded as a single CloudWatch Log record.
+        /// The format of the message must be:
+        /// +----------------------+------------------------+-----------------------+
+        /// | Frame Type - 4 bytes | Length (len) - 4 bytes | Message - 'len' bytes |
+        /// +----------------------+------------------------+-----------------------+
+        /// The first 4 bytes are the frame type. For logs this is always 0xa55a0001.
+        /// The second 4 bytes are the length of the message.
+        /// The remaining bytes are the message itself. Byte order is big-endian.
+        /// </summary>
+        private class FileDescriptorLogStream : Stream
+        {
+            const uint FRAME_TYPE = 0xa55a0001;
+            Stream _fileDescriptorStream;
+            byte[] _frameTypeBytes;
+
+            public FileDescriptorLogStream(string fileDescriptorId)
+            {
+                SafeFileHandle handle = new SafeFileHandle(new IntPtr(int.Parse(fileDescriptorId)), false);
+                _fileDescriptorStream = new FileStream(handle, FileAccess.Write);
```

Comment:
is it possible two lambda executions trying to write same file or `fileDescriptorId` is always unique per lambda execution?

Reply:
There is one `fileDescriptorId` per Lambda execution environment. The parent classes `GetWriter` makes sure we only use a single writer for a unique `fileDescriptorId`.